local/sdk-sriov/pkg/k8s/deviceplugin/server.go

```diff
@@ -0,0 +1,321 @@
+		hostPathEnv:         config.HostPathEnv,
+		devices:             map[string]*device{},
+		updateDevicesCh:     make(chan bool, 1),
+	}
+}
+
+// StartServer creates a new SR-IOV forwarder device plugin server and starts it
+func StartServer(ctx context.Context, config *ServerConfig, manager K8sManager) error {
+	logEntry := log.Entry(ctx).WithField("devicePluginServer", "StartServer")
+
+	s := newDevicePluginServer(ctx, config)
+
+	logEntry.Info("get resource lister client")
+	resourceListerClient, err := manager.GetPodResourcesListerClient(ctx)
+	if err != nil {
+		logEntry.Error("failed to get resource lister client")
+		return err
+	}
+	s.resourceListerClient = resourceListerClient
+
+	logEntry.Info("starting server")
+	socket, err := manager.StartDeviceServer(s.ctx, s)
+	if err != nil {
+		logEntry.Error("error starting server")
+		return err
+	}
+
+	logEntry.Info("registering server")
+	if err := manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     socket,
+		ResourceName: s.resourceName,
+	}); err != nil {
+		logEntry.Error("error registering server")
+		return err
+	}
+
+	if err := s.monitorKubeletRestart(manager, socket); err != nil {
+		logEntry.Warnf("error monitoring kubelet restart: %+v", err)
+	}
+
+	return nil
+}
+
+func (s *devicePluginServer) monitorKubeletRestart(manager K8sManager, socket string) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "monitorKubeletRestart")
+
+	resetCh, err := manager.MonitorKubeletRestart(s.ctx)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		logEntry.Infof("start monitoring kubelet restart")
+		defer logEntry.Infof("stop monitoring kubelet restart")
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case _, ok := <-resetCh:
+				if !ok {
+					return
+				}
+				logEntry.Info("re registering server")
+				if err = manager.RegisterDeviceServer(s.ctx, &pluginapi.RegisterRequest{
+					Version:      pluginapi.Version,
+					Endpoint:     socket,
+					ResourceName: s.resourceName,
+				}); err != nil {
+					logEntry.Errorf("error re registering server: %+v", err)
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *devicePluginServer) GetDevicePluginOptions(_ context.Context, _ *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	return &pluginapi.DevicePluginOptions{}, nil
+}
+
+func (s *devicePluginServer) ListAndWatch(_ *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	logEntry := log.Entry(s.ctx).WithField("devicePluginServer", "ListAndWatch")
+
+	for {
+		resp, err := s.resourceListerClient.List(s.ctx, &podresources.ListPodResourcesRequest{})
+		if err != nil {
+			logEntry.Errorf("resourceListerClient unavailable: %+v", err)
+			return err
+		}
+
+		s.updateDevices(s.respToDeviceIds(resp))
+
+		if err := server.Send(s.listAndWatchResponse()); err != nil {
+			logEntry.Errorf("server unavailable: %+v", err)
+			return err
+		}
```

Comment:
Probably yes, but I am not sure. Guess we first need to see it at work.

Reply:
Do we have some retry logic in monorepo for dp?

Reply:
Monorepo doesn't stop trying: https://github.com/networkservicemesh/networkservicemesh/blob/14700b7a14a4a722510b12da7ecf28fa369d137f/k8s/cmd/nsmdp/nsmdp.go#L304
But actually I don't think that it is true. If `server.Send(...)` fails it should mean that there is some connection that has been closed and so there is nothing we can do here more.
Actually, every time we reregister the server, kubelet calls `dps.ListAndWatch(...)` so it is a retry logic if we need to retry in case of the connection reopening.

Reply:
I think we should add retries because with the current approach we can face unexpected behavior if something happened with k8s server. For example, if k8s server is busy, then as I understand the device plugin starts work incorrectly.